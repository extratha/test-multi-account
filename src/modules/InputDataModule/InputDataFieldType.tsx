import { Grid, Stack, styled, Typography } from "@mui/material";
import { useTranslations } from "next-intl";

import FormAutocomplete from "@/components/Form/FormAutocomplete";
import FormNumberInput from "@/components/Form/FormNumberInput";
import { CONFIG_FIELD_TYPES } from "@/constant/constant";
import { Field } from "@/types/interpretInputDataConfig";

interface InputDataFieldTypeProps {
  field: Field;
}
// TODO: Unit test

const InputDataFieldWrapper = styled(Grid)({
  padding: "1rem",
});

const TypoUnit = styled(Typography)(({ theme }) => ({
  wordWrap: "break-word",
  color: theme.palette.grey[600],
  margin: "auto 0",
}));

const InputDataFieldType = ({ field }: InputDataFieldTypeProps) => {
  const tAi = useTranslations("AiInterpret");

  const getPlaceholder = () => {
    if (field.key === "gender") {
      return tAi("th.placeholder.gender");
    }
    return "";
  };

  const getDropdownOptions = () => {
    const dropdown = field.dropdownValue || [];
    return dropdown.map((value) => ({
      label: value,
      value,
    }));
  };

  const displayUnit = (field: Field) => {
    if (field.key === "age") return tAi("field.yearsOld");
    return field.unit;
  };

  return (
    <InputDataFieldWrapper container>
      <Grid item xs={8}>
        <Stack direction="row">
          <Typography variant="titleMediumSemiBold">{tAi(`th.field.${field.key}`)}</Typography>
          {field.required && (
            <Typography ml={0.5} variant="titleLargeSemiBold" color="error.light">
              *
            </Typography>
          )}
        </Stack>
        <Typography variant="titleMedium" color="grey.600">{`(${tAi(`en.field.${field.key}`)})`}</Typography>
      </Grid>
      <Grid item xs={4}>
        <Grid container>
          <Grid xs={10} pr={2}>
            {field.fieldType === CONFIG_FIELD_TYPES.DROPDOWN && (
              <FormAutocomplete
                name={field.key}
                label={tAi(`th.field.${field.key}`)}
                options={getDropdownOptions()}
                required={field.required}
                placeholder={getPlaceholder()}
              />
            )}
            {field.fieldType === CONFIG_FIELD_TYPES.NUMBER && (
              <FormNumberInput name={field.key} label={tAi(`th.field.${field.key}`)} required={field.required} />
            )}
          </Grid>
          <Grid xs={2}>
            <TypoUnit variant="bodyLarge">{displayUnit(field)}</TypoUnit>
          </Grid>
        </Grid>
      </Grid>
    </InputDataFieldWrapper>
  );
};

export default InputDataFieldType;
