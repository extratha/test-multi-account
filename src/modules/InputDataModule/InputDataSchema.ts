import { useMemo } from "react";
import * as Yup from "yup";

import { InputGroupConfig } from "@/types/model.ui";
import useTranslation from "@/locales/useLocale";

export const useInputDataFieldYupSchema = (config: InputGroupConfig[]) => {
  const { translation } = useTranslation();

  const fieldConfig = useMemo(() => {
    const shape: Record<string, Yup.AnySchema> = {};

    config.forEach((group) => {
      group.data.forEach((field) => {
        let schema: Yup.AnySchema;

        const error = {
          message: {
            require: translation("Common.validation.require"),
            selectRequire: translation("Common.validation.selectRequire"),
            range: `${translation("Common.validation.valueMustBeInRange")} ${field.minValue}-${
              field.maxValue
            } ${translation("Common.validation.only")}`,
            invalidInput: translation("Common.validation.invalidInput"),
            length: `${translation("Common.validation.valueMustNotExceed")} ${field.maxLength} ${translation(
              "Common.validation.characters"
            )} `,
          },
        };
        switch (field.fieldType) {
          case "Number":
            schema = Yup.number()
              .typeError(error.message.invalidInput)
              .transform((value, originalValue) => (originalValue === "" ? undefined : value));

            if (field.required) {
              schema = schema.required(error.message.require);
            }
            if (field.minValue) {
              schema = (schema as Yup.NumberSchema).min(Number(field.minValue), error.message.range);
            }
            if (field.maxValue) {
              schema = (schema as Yup.NumberSchema).max(Number(field.maxValue), error.message.range);
            }
            break;
          case "Dropdown":
            schema = Yup.string().nullable();
            if (field.required) {
              schema = schema.required(error.message.selectRequire);
            }
            if (field.dropdownValue) {
              schema = schema.oneOf(field.dropdownValue, error.message.invalidInput);
            }
            break;
          default:
            schema = Yup.string();
            if (field.required) {
              schema = schema.required(error.message.require);
            }
            if (field.maxLength) {
              schema = (schema as Yup.StringSchema).max(Number(field.maxLength), error.message.length);
            }
            break;
        }

        shape[field.key] = schema;
      });
    });

    return Yup.object().shape(shape);
  }, [config.length]);

  return fieldConfig;
};
