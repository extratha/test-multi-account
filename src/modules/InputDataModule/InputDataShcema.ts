import { FieldConfig } from "@/types/interpretInputDataConfig";
import * as Yup from "yup";

export const inputDataFieldYupSchema = (fieldConfig: FieldConfig) => {
  const shape: Record<string, Yup.AnySchema> = {};

  fieldConfig.forEach((group) => {
    group.data.forEach((field) => {
      let schema: Yup.AnySchema = Yup.string();

      switch (field.fieldType) {
        case "Number":
          schema = Yup.number();
          if (field.required) {
            schema = schema.required("This field is required");
          }
          if (field.minLength) {
            schema = (schema as Yup.NumberSchema).min(Number(field.minLength), `Minimum value is ${field.minLength}`);
          }
          if (field.maxLength) {
            schema = (schema as Yup.NumberSchema).max(Number(field.maxLength), `Maximum value is ${field.maxLength}`);
          }
          break;
        case "Dropdown":
          schema = Yup.string();
          if (field.required) {
            schema = schema.required("This field is required");
          }
          if (field.dropdownValue) {
            schema = schema.oneOf(field.dropdownValue, `Invalid value`);
          }
          break;
        default:
          schema = Yup.string();
          if (field.required) {
            schema = schema.required("This field is required");
          }
          if (field.minLength) {
            schema = (schema as Yup.StringSchema).min(Number(field.minLength), `Minimum length is ${field.minLength}`);
          }
          if (field.maxLength) {
            schema = (schema as Yup.StringSchema).max(Number(field.maxLength), `Maximum length is ${field.maxLength}`);
          }
          break;
      }

      shape[field.key] = schema;
    });
  });

  return Yup.object().shape(shape);
};
