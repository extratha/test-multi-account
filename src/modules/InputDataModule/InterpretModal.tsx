import { Button, Dialog as MuiDialog, Stack, styled, Typography } from "@mui/material";
import Image from "next/image";

import { IconErrorCircle, ImageLoadingStack } from "@/assets";
import { INTERPRET_STATUS } from "@/constant";
import useTranslation from "@/locales/useLocale";

const Dialog = styled(MuiDialog)(({ theme }) => ({
  "& .MuiPaper-root": {
    backgroundColor: theme.palette.common.transparent,
    boxShadow: "none",
  },
}));

const Content = styled(Stack)(({ theme }) => ({
  alignItems: "center",
  color: theme.palette.common.white,
}));

const StatusImage = styled(Image)({
  width: "100px",
  height: "100px",
});

const StatusMessage = styled(Typography)({
  marginTop: "18px",
  whiteSpace: "pre-line",
});

const ButtonAction = styled(Button)(({ theme }) => ({
  width: "150px",
  height: "46px",
  marginTop: "36px",
  backgroundColor: theme.palette.common.transparent,
  color: theme.palette.common.white,
  border: `2px solid ${theme.palette.common.white}`,
}));

export interface InterpretModalProps {
  status: string;
  onClose?: () => void;
}

const InterpretModal = ({ status, onClose }: InterpretModalProps) => {
  const { translation } = useTranslation();

  return (
    <Dialog open>
      <Content>
        {status === INTERPRET_STATUS.PENDING && (
          <>
            <StatusImage data-testid="interpret-image-pending" alt="" src={ImageLoadingStack} />
            <StatusMessage variant="bodyBold">{translation("AiInterpret.text.interpreting")}</StatusMessage>
            <ButtonAction data-testid="modal-ok-button" onClick={onClose}>
              {translation("AiInterpret.button.cancel")}
            </ButtonAction>
          </>
        )}
        {status === INTERPRET_STATUS.FAILED && (
          <>
            <IconErrorCircle data-testid="interpret-image-failed" />
            <StatusMessage variant="bodyBold" textAlign="center">
              {translation("AiInterpret.text.failedInterpret")}
            </StatusMessage>
            <ButtonAction data-testid="modal-retry-button" onClick={onClose}>
              {translation("AiInterpret.button.retry")}
            </ButtonAction>
          </>
        )}
      </Content>
    </Dialog>
  );
};

export default InterpretModal;
