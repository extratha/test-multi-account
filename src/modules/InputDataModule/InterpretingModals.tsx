import { IconErrorCircle, ImageLoadingStack } from "@/assets";
import { INTERPRET_STATUS } from "@/constant/constant";
import useTranslation from "@/locales/useLocale";
import { BaseModalProps } from "@/store";
import { Button, Stack, styled, Typography } from "@mui/material";
import Image from "next/image";

const ModalStackWrapper = styled(Stack)(({ theme }) => ({
  alignItems: "center",
  color: theme.palette.background.paper,
}));

const StatusImageInModal = styled(Image)({
  width: "100px",
  height: "100px",
});

const StatusMessage = styled(Typography)({
  marginTop: "18px",
  whiteSpace: "pre-line",
});

const ButtonInModal = styled(Button)(({ theme }) => ({
  width: "150px",
  height: "46px",
  marginTop: "36px",
  backgroundColor: "transparent",
  color: theme.palette.background.paper,
  border: `2px solid ${theme.palette.background.paper}`,
  fontWeight: 600,
}));

export interface InterpretingModalsProps extends BaseModalProps {
  interpretStatus: string;
}

const InterpretingModals = (props: InterpretingModalsProps) => {
  const { closeModal, interpretStatus } = props;
  const { translation } = useTranslation();

  return (
    <ModalStackWrapper>
      {interpretStatus === INTERPRET_STATUS.PENDING && (
        <>
          <StatusImageInModal data-testid="interpret-image-pending" alt="" src={ImageLoadingStack} />
          <StatusMessage variant="bodyBold">{translation("AiInterpret.text.interpreting")}</StatusMessage>
          <ButtonInModal data-testid="modal-ok-button" onClick={closeModal}>
            {translation("AiInterpret.button.ok")}
          </ButtonInModal>
        </>
      )}
      {interpretStatus === INTERPRET_STATUS.FAILED && (
        <>
          <IconErrorCircle data-testid="interpret-image-failed" />
          <StatusMessage variant="bodyBold" textAlign="center">
            {translation("AiInterpret.text.failedInterpret")}
          </StatusMessage>
          <ButtonInModal data-testid="modal-retry-button" onClick={closeModal}>
            {translation("AiInterpret.button.retry")}
          </ButtonInModal>
        </>
      )}
    </ModalStackWrapper>
  );
};

export default InterpretingModals;
