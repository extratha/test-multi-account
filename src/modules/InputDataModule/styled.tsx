import { CUSTOM_COLORS } from "@/config/config-mui/theme/colors";
import { Button, Stack, styled } from "@mui/material";

export const CommonButton = styled(Button)(({ theme }) => ({
  padding: 10,
  height: 40,
  background: theme.palette.background.paper,
  borderRadius: "10px",
  border: `1px solid ${theme.palette.grey[500]}`,
  width: "fit-content",
  color: CUSTOM_COLORS.buttonText,
}));

export const ContentContainerWrapper = styled(Stack)(({ theme }) => ({
  backgroundColor: theme.palette.background.paper,
  borderRadius: "20px",
  maxWidth: "1080px",
  width: "100%",
  height: "100%",
  padding: "2rem",
  margin: "auto",
  overflowY: "auto",
}));
