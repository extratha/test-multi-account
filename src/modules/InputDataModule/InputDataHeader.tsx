import { Button, Stack, styled, Typography } from "@mui/material";

import { IconImportExampleData, IconSparkle, IconSparkleDisabled } from "@/assets";
import useTranslation from "@/locales/useLocale";
import { ButtonInterpretDataStyled } from "../ExampleDataList/styled";

const CommonButton = styled(Button)(({ theme }) => ({
  height: 40,
  padding: "10px 16px",
  borderRadius: "8px",
  color: theme.palette.text.hight,
  border: `1px solid ${theme.palette.background.border}`,
  background: theme.palette.background.paper,
  "&:hover": {
    border: `1px solid ${theme.palette.background.border}`,
  },
}));

export interface InputDataHeaderProps {
  isDisableSubmit: boolean;
  modelVersion?: string;
  onClickUseExampleData: () => void;
  onSubmit: () => void;
}

const InputDataHeader = (props: InputDataHeaderProps) => {
  const { isDisableSubmit, onSubmit, onClickUseExampleData, modelVersion } = props;
  const { translation } = useTranslation();
  return (
    <Stack mt={3}>
      <Stack direction="row">
        <Stack justifyContent="center">
          <Typography data-testid="page-title" variant="headerSemiBold">
            {translation("AiInterpret.pages.tryInputData")}
          </Typography>
        </Stack>
        <Stack ml="auto">
          <Stack direction="row" spacing={2} justifyContent="end">
            <CommonButton
              variant="outlined"
              startIcon={<IconImportExampleData />}
              data-testid="use-example-data-button"
              onClick={onClickUseExampleData}
            >
              {translation("AiInterpret.button.useExampleData")}
            </CommonButton>
            <ButtonInterpretDataStyled
              variant="contained"
              startIcon={isDisableSubmit ? <IconSparkleDisabled /> : <IconSparkle />}
              disabled={isDisableSubmit}
              data-testid="submit-interpret-button"
              onClick={onSubmit}
            >
              {translation("AiInterpret.button.interpretData")}
            </ButtonInterpretDataStyled>
          </Stack>
          <Stack direction="row" mt={2}>
            <Stack direction="row" spacing={1}>
              <Typography variant="labelExtraSmall">{translation("AiInterpret.text.myCredit")}</Typography>
              <Typography variant="labelExtraSmallBold">{translation("AiInterpret.text.unlimited")}</Typography>
            </Stack>
            {modelVersion && (
              <Stack direction="row" ml={2} spacing={1}>
                <Typography variant="labelExtraSmall">{translation("AiInterpret.field.modelVersion")}</Typography>
                <Typography variant="labelExtraSmallBold">{modelVersion}</Typography>
              </Stack>
            )}
          </Stack>
        </Stack>
      </Stack>
    </Stack>
  );
};

export default InputDataHeader;
