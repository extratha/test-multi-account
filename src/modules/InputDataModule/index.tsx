"use client";

import { yupResolver } from "@hookform/resolvers/yup";
import { Button, Divider, Paper, Stack, styled, Typography } from "@mui/material";
import { useRouter, useSearchParams } from "next/navigation";
import { useEffect, useMemo, useState } from "react";
import { FormProvider, useForm } from "react-hook-form";

import { getLabInterpretsByTransactionId, submitLabInterprets } from "@/api/api";
import { IconArrowLeft } from "@/assets";
import { INTERPRET_STATUS } from "@/constant/constant";
import { webPaths } from "@/constant/webPaths";
import { useGetLabExampleId } from "@/hooks/useApi";
import useTranslation from "@/locales/useLocale";
import { usePageLoadingStore } from "@/store";
import useModal from "@/store/modal";
import { InterpretResult } from "@/types/model.api";
import { InputDataConfig, InputGroupConfigResult } from "@/types/model.ui";
import { getLabInterpretFieldsConfig } from "@/utils/firebase";
import { mapInputDataToSubmitInterprets } from "@/utils/mapper";
import InputDataFieldType from "./InputDataFieldType";
import InputDataHeader from "./InputDataHeader";
import { useInputDataFieldYupSchema } from "./InputDataSchema";
import InterpretModals from "./InterpretingModals";

export type FormInputDataValuesType = Record<string, unknown>;

interface DefaultValues {
  [key: string]: string;
}

const CommonButton = styled(Button)(({ theme }) => ({
  height: 40,
  color: theme.palette.text.hight,
  borderRadius: "10px",
  border: `1px solid ${theme.palette.blueGrey[100]}`,
  background: theme.palette.background.paper,
  "&:hover": {
    border: `1px solid ${theme.palette.blueGrey[100]}`,
  },
}));

const ContentContainer = styled(Stack)(({ theme }) => ({
  flex: 1,
  width: "100%",
  padding: "24px",
  backgroundColor: theme.palette.background.grayLight,
}));

const ContentContainerWrapper = styled(Paper)({
  flex: 1,
  width: "100%",
  maxWidth: "1080px",
  padding: "32px",
  borderRadius: "20px",
  margin: "0px auto",
});

const InputDataGroupContainer = styled(Stack)(({ theme }) => ({
  margin: "2rem 0",
  borderRadius: "8px",
  border: `1px solid ${theme.palette.background.borderLight}`,
}));

const InputDataGroupHeader = styled(Stack)(({ theme }) => ({
  padding: "24px 24px 12px",
  borderRadius: "8px 8px 0px 0px",
  backgroundColor: theme.palette.surfaceGray.lowest,
}));

const InputDataGroupContent = styled(Stack)({
  padding: "16px",
});

const MAX_INTERVAL = 60000;
const INTERVAL_DELAY = 5000;

const InputDataModule = () => {
  const router = useRouter();
  const searchParams = useSearchParams();
  const { translation } = useTranslation();
  const { openModal, closeModal } = useModal();
  const { isPageLoading, setPageLoading } = usePageLoadingStore();

  const interpretId = searchParams.get("exampleId");

  const [inputGroupConfigs, setInputGroupConfigs] = useState<InputGroupConfigResult[]>([]);

  const { data, isLoading: isGetLabExampleIdLoading } = useGetLabExampleId(interpretId || "");
  const inputData = data?.data?.inputData || [];
  const modelVersion = data?.data?.aiModelVersion || "";

  const isLoading = isGetLabExampleIdLoading || isPageLoading;

  const validateSchema = useInputDataFieldYupSchema(inputGroupConfigs);
  const methods = useForm<FormInputDataValuesType>({
    resolver: yupResolver(validateSchema),
    defaultValues: {},
    mode: "onChange",
  });

  const { formState, setValue, trigger } = methods;
  const isDisableInterpretButton = isGetLabExampleIdLoading || !formState.isValid;

  const defaultInputData: DefaultValues = useMemo(() => {
    const result: DefaultValues = {};

    if (inputData.length === 0) {
      return result;
    }

    inputData.forEach((group) => {
      group.data.forEach((item) => {
        result[item.key] = item.value;
      });
    });
    return result;
  }, [inputData.length]);

  const fetchConfigData = async () => {
    setPageLoading(true);
    const remoteConfig = await getLabInterpretFieldsConfig();
    setInputGroupConfigs(remoteConfig);
    setPageLoading(false);
  };

  const handleClickBackButton = () => {
    router.replace(webPaths.aiInterpret.tryExampleData);
  };

  const handleClickUseExampleData = () => {
    router.replace(webPaths.aiInterpret.tryExampleData);
  };

  const handleSetDefaultValue = () => {
    const fieldNames = Object.keys(defaultInputData);

    const setFieldValues = () => {
      if (defaultInputData && fieldNames.length > 0) {
        fieldNames.map((fieldName) => {
          setValue(fieldName, defaultInputData[fieldName], { shouldValidate: true });
          trigger(fieldName);
        });
      }
    };

    setFieldValues();
  };

  const waitTimer = (millisecond: number) => {
    return new Promise<boolean>((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, millisecond);
    });
  };

  const fetchInterpretResult = async (transactionID: string, startTime: number): Promise<InterpretResult> => {
    if (Date.now() - startTime > MAX_INTERVAL) {
      throw new Error("fetchInterpretResult: timeout");
    }

    const { data } = await getLabInterpretsByTransactionId(transactionID);

    if (data.status === INTERPRET_STATUS.SUCCESS) {
      return data;
    } else {
      await waitTimer(INTERVAL_DELAY);
      return fetchInterpretResult(transactionID, startTime);
    }
  };

  const onSubmit = async (formValues: FormInputDataValuesType) => {
    try {
      openModal((props) => <InterpretModals {...props} interpretStatus={INTERPRET_STATUS.PENDING} />, false);
      const response = await submitLabInterprets(mapInputDataToSubmitInterprets(formValues, inputGroupConfigs));
      const aiResult = await fetchInterpretResult(response.data.transactionID, Date.now());

      router.replace(`${webPaths.aiInterpret.aiInterpretResult}?transactionId=${aiResult.id}`);
      closeModal();
    } catch (error) {
      openModal((props) => <InterpretModals {...props} interpretStatus={INTERPRET_STATUS.FAILED} />, false);
    }
  };

  useEffect(() => {
    fetchConfigData();
  }, []);

  useEffect(() => {
    handleSetDefaultValue();
  }, [defaultInputData]);

  useEffect(() => {
    if (inputData.length > 0 && defaultInputData) {
      trigger();
    }
  }, [inputData.length, isLoading]);

  return (
    <>
      {!isLoading && (
        <ContentContainer>
          <ContentContainerWrapper>
            <Stack mb="16px" alignItems="flex-start">
              <CommonButton
                variant="outlined"
                startIcon={<IconArrowLeft />}
                data-testid="back-button"
                onClick={handleClickBackButton}
              >
                {translation("AiInterpret.button.backToMain")}
              </CommonButton>
            </Stack>
            <Divider />
            <InputDataHeader
              isDisableSubmit={isDisableInterpretButton}
              modelVersion={modelVersion}
              onSubmit={methods.handleSubmit(onSubmit)}
              onClickUseExampleData={handleClickUseExampleData}
            />
            <FormProvider {...methods}>
              <form onSubmit={methods.handleSubmit(onSubmit)}>
                {inputGroupConfigs.map((group, groupIndex) => (
                  <InputDataGroupContainer key={groupIndex}>
                    <InputDataGroupHeader>
                      <Typography variant="titleBold">
                        {translation(`AiInterpret.th.groupName.${group.groupName}`)}
                      </Typography>
                      <Typography variant="bodySmall" marginTop="4px">{`(${translation(
                        `AiInterpret.en.groupName.${group.groupName}`
                      )})`}</Typography>
                    </InputDataGroupHeader>
                    <InputDataGroupContent>
                      {group.data.map((field: InputDataConfig, fieldIndex: number) => (
                        <InputDataFieldType key={fieldIndex} field={field} />
                      ))}
                    </InputDataGroupContent>
                  </InputDataGroupContainer>
                ))}
              </form>
            </FormProvider>
          </ContentContainerWrapper>
        </ContentContainer>
      )}
    </>
  );
};

export default InputDataModule;
