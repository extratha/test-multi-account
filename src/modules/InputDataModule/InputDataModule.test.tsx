import "firebase/remote-config";
import { useRouter } from "next/navigation";
import InputDataModule from ".";
import { flushPromise, render, screen, userEvent } from "../../__tests__/testUtils";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

describe("InputDataModule", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderInputDataModule = async () => {
    const view = render(<InputDataModule />);
    await flushPromise();
    return view;
  };

  it("should render correctly", async () => {
    const { asFragment } = await renderInputDataModule();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render page title", async () => {
    await renderInputDataModule();
    const pageTitle = screen.getByTestId("page-title");
    expect(pageTitle).toBeInTheDocument();
  });

  it("call router back when click back", async () => {
    const mockBack = jest.fn();
    const mockUseRouter = useRouter as jest.Mock;
    mockUseRouter.mockReturnValue({
      back: mockBack,
    });
    await renderInputDataModule();

    const backButton = await screen.findByTestId("back-button");
    await userEvent.click(backButton);

    expect(mockBack).toHaveBeenCalledTimes(1);
  });
});
