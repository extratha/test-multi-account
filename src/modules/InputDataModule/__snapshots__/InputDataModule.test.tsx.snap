// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDataModule should render correctly 1`] = `<DocumentFragment />`;
